libimage/runtime.go

```diff
@@ -414,79 +441,65 @@ func (r *Runtime) lookupImageInLocalStorage(name, candidate string, options *Loo
+		return nil, "", fmt.Errorf("%s: %w", originalName, storage.ErrImageUnknown)
 	}
 
-	digested, isDigested := named.(reference.Digested)
+	// In case of a digested reference, we strip off the digest and require
+	// any image matching the repo/tag to also match the specified digest.
+	var requiredDigest digest.Digest
+	digested, isDigested := possiblyUnqualifiedNamedReference.(reference.Digested)
 	if isDigested {
-		logrus.Debug("Looking for image with matching recorded digests")
-		digest := digested.Digest()
-		for _, image := range allImages {
-			for _, d := range image.Digests() {
-				if d != digest {
-					continue
-				}
-				// Also make sure that the matching image fits all criteria (e.g., manifest list).
-				if _, err := r.lookupImageInLocalStorage(name, image.ID(), options); err != nil {
-					return nil, "", err
-				}
-				return image, name, nil
-
-			}
-		}
-		return nil, "", fmt.Errorf("%s: %w", name, storage.ErrImageUnknown)
+		requiredDigest = digested.Digest()
+		possiblyUnqualifiedNamedReference = reference.TrimNamed(possiblyUnqualifiedNamedReference)
+		name = possiblyUnqualifiedNamedReference.String()
 	}
 
 	if !shortnames.IsShortName(name) {
-		return nil, "", fmt.Errorf("%s: %w", name, storage.ErrImageUnknown)
+		return nil, "", fmt.Errorf("%s: %w", originalName, storage.ErrImageUnknown)
+	}
+
+	// Docker compat: make sure to add the "latest" tag if needed.  The tag
+	// will be ignored if we're looking for a digest match.
+	possiblyUnqualifiedNamedReference = reference.TagNameOnly(possiblyUnqualifiedNamedReference)
+	namedTagged, isNamedTagged := possiblyUnqualifiedNamedReference.(reference.NamedTagged)
+	if !isNamedTagged {
+		// NOTE: this should never happen since we already stripped off
+		// the digest.
+		return nil, "", fmt.Errorf("%s: %w (could not cast to tagged)", originalName, storage.ErrImageUnknown)
 	}
 
-	named = reference.TagNameOnly(named) // Make sure to add ":latest" if needed
-	namedTagged, isNammedTagged := named.(reference.NamedTagged)
-	if !isNammedTagged {
-		// NOTE: this should never happen since we already know it's
-		// not a digested reference.
-		return nil, "", fmt.Errorf("%s: %w (could not cast to tagged)", name, storage.ErrImageUnknown)
+	allImages, err := r.ListImages(context.Background(), nil, nil)
+	if err != nil {
+		return nil, "", err
 	}
 
 	for _, image := range allImages {
-		named, err := image.inRepoTags(namedTagged)
+		named, err := image.inRepoTags(namedTagged, isDigested)
 		if err != nil {
 			return nil, "", err
 		}
 		if named == nil {
 			continue
 		}
-		img, err := r.lookupImageInLocalStorage(name, named.String(), options)
+		img, err := r.lookupImageInLocalStorage(name, named.String(), named, options)
 		if err != nil {
 			return nil, "", err
 		}
 		if img != nil {
-			return img, named.String(), err
+			if isDigested && !img.hasDigest(requiredDigest.String()) {
+				continue
+			}
+			return img, named.String(), nil
```

Comment:
AFAICS this now returns _repo_`:`_digestOrSomeAplicableTag` for digested inputs, which seems inaccurate; I’d expect the return value to be a _repo_`@`_digest_.

I’m not 100% sure what _repo_ value we should be returning — the previous version of the digest path returned the input unchanged, even if it was a short name. It seems to me that it should return a fully-qualified name; in that case we probably need `inRepoTags` to return a name with the matched fully-qualified repo, and to construct a repo@digest right here.

---

Consider changing `inRepoTags` not to require a `NamedTagged` input, but just a `Named`, and to pass the tag value separately (and with `""` meaning “ignore tag”, we wouldn’t need the separate `ignoreTag` parameter). Then the original `possiblyUnqualifiedNamedReference` could be passed to `inRepoTags` unchanged.

Reply:
> AFAICS this now returns  …

It can return a precise repo@digest match if that’s how the image was pulled; for digest references to images pulled by tag, it will return a _repo_`:`_someApplicableTag_.

Reply:
Oh man, I am so glad you're finding such corner cases.  I will add proper tests for `LookupImage` to make sure we won't regress.